src/ring_theory/witt_vector/truncated.lean

```diff
@@ -0,0 +1,246 @@
+  simpa only [coeff_out] using h ↑i
+end
+
+end truncated_witt_vector
+
+namespace witt_vector
+
+variables {p} (n)
+
+section
+
+local attribute [semireducible] witt_vector
+
+/-- `truncate_fun n x` uses the first `n` entries of `x` to construct a `truncated_witt_vector`,
+which has the same base `p` as `x`.
+This function is bundled into a ring homomorphism in `witt_vector.truncate` -/
+def truncate_fun (x : 𝕎 R) : truncated_witt_vector p n R :=
+truncated_witt_vector.mk p $ λ i, x.coeff i
+
+end
+
+variables {n}
+
+@[simp] lemma coeff_truncate_fun (x : 𝕎 R) (i : fin n) :
+  (truncate_fun n x).coeff i = x.coeff i :=
+by rw [truncate_fun, truncated_witt_vector.coeff_mk]
+
+variable [comm_ring R]
+
+@[simp] lemma out_truncate_fun (x : 𝕎 R) :
+  (truncate_fun n x).out = init n x :=
+begin
+  ext i,
+  dsimp [truncated_witt_vector.out, init, select],
+  split_ifs with hi, swap, refl,
+  rw [coeff_truncate_fun, fin.coe_mk],
+end
+
+end witt_vector
+
+namespace truncated_witt_vector
+
+variable [comm_ring R]
+
+@[simp] lemma truncate_fun_out (x : truncated_witt_vector p n R) :
+  x.out.truncate_fun n = x :=
+by simp only [witt_vector.truncate_fun, coeff_out, mk_coeff]
+
+end truncated_witt_vector
+
+namespace truncated_witt_vector
+open witt_vector
+variables (p n R)
+variable [comm_ring R]
+
+include hp
+
+instance : has_zero (truncated_witt_vector p n R) :=
+⟨truncate_fun n 0⟩
+
+instance : has_one (truncated_witt_vector p n R) :=
+⟨truncate_fun n 1⟩
+
+instance : has_add (truncated_witt_vector p n R) :=
+⟨λ x y, truncate_fun n (x.out + y.out)⟩
+
+instance : has_mul (truncated_witt_vector p n R) :=
+⟨λ x y, truncate_fun n (x.out * y.out)⟩
+
+instance : has_neg (truncated_witt_vector p n R) :=
+⟨λ x, truncate_fun n (- x.out)⟩
+
+@[simp] lemma coeff_zero (i : fin n) :
+  (0 : truncated_witt_vector p n R).coeff i = 0 :=
+begin
+  show coeff i (truncate_fun _ 0 : truncated_witt_vector p n R) = 0,
+  rw [coeff_truncate_fun, witt_vector.zero_coeff],
+end
+
+end truncated_witt_vector
+
+/-- A macro tactic used to prove that `truncate_fun` respects ring operations. -/
+meta def tactic.interactive.truncate_fun_tac : tactic unit :=
+`[show _ = truncate_fun n _,
+  apply truncated_witt_vector.out_injective,
+  iterate { rw [out_truncate_fun] },
+  rw init_add <|> rw init_mul <|> rw init_neg]
+
+namespace witt_vector
+
+variables (p n R)
+variable [comm_ring R]
+
+lemma truncate_fun_surjective :
+  surjective (@truncate_fun p n R) :=
+λ x, ⟨x.out, truncated_witt_vector.truncate_fun_out x⟩
+
+include hp
+
+@[simp]
+lemma truncate_fun_zero : truncate_fun n (0 : 𝕎 R) = 0 := rfl
+
+@[simp]
+lemma truncate_fun_one : truncate_fun n (1 : 𝕎 R) = 1 := rfl
+
+variables {p R}
+
+@[simp]
+lemma truncate_fun_add (x y : 𝕎 R) :
+  truncate_fun n (x + y) = truncate_fun n x + truncate_fun n y :=
+by truncate_fun_tac
+
+@[simp]
+lemma truncate_fun_mul (x y : 𝕎 R) :
+  truncate_fun n (x * y) = truncate_fun n x * truncate_fun n y :=
+by truncate_fun_tac
```

Comment:
Can you bundle this into a `ring_hom`? You mention about that _"This function is bundled into a ring homomorphism in `witt_vector.truncate`", but I assume that's part of a follow-up PR?

Reply:
Exactly. You need the function to get the ring structure. So you can't bundle it before that. I thought that a natural point to split the file was once the ring structure was established. The bundled `truncate` is one of the first things in the 2nd half.